File: HPC/NeuralNetwork.cpp
#include <iostream>
#include <algorithm>

#include "NeuralNetwork.h"

#if 0
float NeuralNetwork::forward(std::shared_ptr<Eigen::Tensor<float, 4>> images, std::shared_ptr<Eigen::Tensor<float, 2>> label_tensor) {

	for (const auto& layer: layers) {
		images = layer->forward(images);
	}
	
	return loss_layer.forward(images, label_tensor);
}

void NeuralNetwork::backward(std::shared_ptr<Eigen::Tensor<float, 2>> label_tensor) {
	
	auto error_tensor = loss_layer.backward(label_tensor);

	for (auto it = layers.rbegin(); it != layers.rend(); ++it)
	{
		error_tensor = (*it)->backward(error_tensor);
	}
}

void NeuralNetwork::appendLayer(std::unique_ptr<Layer> layer) {
	if (layer->trainable) {
		layer->setOptimizer(optimizer);
		layer->setInitializer(initializer);
	}

	layers.push_back(std::move(layer));

}

void NeuralNetwork::train(int iterations) {
	for (int i = 0; i < iterations; i++) {
		auto batch = batch_loader.loadBatch(batch_size);

		float loss = forward(batch.second, batch.first);

		std::cout << "(" << i << ")" << " Loss: " << loss << std::endl;
		
		losses.push_back(loss);
		backward(batch.first);
	}
}

std::shared_ptr<Eigen::Tensor<float, 4> const> NeuralNetwork::test(std::shared_ptr<Eigen::Tensor<float, 4> const> test_data) {
	for (const auto& layer: layers) {
		test_data = layer->forward(test_data);
	}
	return test_data;
}
#endif

File: cnn_python/NeuralNetwork.py
import copy

class NeuralNetwork:

    def __init__(self, optimizer, weights_initializer, bias_initializer):
        self.optimizer = optimizer
        self.weights_initializer = weights_initializer
        self.bias_initializer = bias_initializer
        self.loss = []
        self.layers = []
        self.data_layer = None
        self.loss_layer = None


    def forward(self):
        input_tensor, self.label_tensor = self.data_layer.next()
        for layer in self.layers:
            input_tensor = layer.forward(input_tensor)
        return self.loss_layer.forward(input_tensor, self.label_tensor)


    def backward(self):
        error_tensor = self.loss_layer.backward(self.label_tensor)
        for layer in reversed(self.layers):
            # print(error_tensor)
            error_tensor = layer.backward(error_tensor)


    def append_layer(self, layer):
        if layer.trainable:
            layer.optimizer = copy.deepcopy(self.optimizer)
            layer.initialize(self.weights_initializer, self.bias_initializer)
        self.layers.append(layer)


    def train(self, iterations):
        for _ in range(iterations):
            self.loss.append(self.forward())
            if _ % 25 == 0:
                print(self.loss[-1])
            self.backward()

    
    def test(self, input_tensor):
        for layer in self.layers:
            input_tensor = layer.forward(input_tensor)
        return input_tensor
    
